scripts/cybershake/cybershake_progress.py

```diff
@@ -0,0 +1,364 @@
+        c.execute(
+            f"SELECT COUNT(*) FROM state WHERE proc_type=? and status={const.Status['completed'].value}",
+            (proc_type_int,),
+        )
+    else:
+        c.execute(
+            f"SELECT COUNT(*) FROM state WHERE proc_type=? and status={const.Status['completed'].value} and run_name LIKE ?",
+            (proc_type_int, fault_name + "_REL%"),
+        )
+
+    completed_r_count = c.fetchone()[0]
+    conn.close()
+    return completed_r_count
+
+
+def get_faults_and_r_count(cybershake_list: str):
+    """Gets the fault names and number of realisations from a cybershake fault list."""
+    with open(cybershake_list, "r") as f:
+        lines = f.readlines()
+
+    fault_names, r_counts = [], []
+    for ix, line in enumerate(lines):
+        line_l = line.split(" ")
+        fault_names.append(line_l[0].strip())
+
+        try:
+            r_counts.append(int(line_l[1].strip().rstrip("RrNnKk")))
+        except ValueError:
+            print(
+                "Failed to read line {} of cybershake list. Need to know"
+                "the number of realisations for each fault, quitting!".format(ix + 1)
+            )
+            # Raise the exception so execution stops
+            raise
+
+    return np.asarray(fault_names), np.asarray(r_counts)
+
+
+def get_new_progress_df(root_dir, runs_dir, faults, fault_names, r_counts):
+    """Gets a new progress dataframe, runs the full estimation + collects
+    all actual core hours.
+    """
+    # Run the estimation
+    est_args = argparse.Namespace(
+        vms_dir=sim_struct.get_VM_dir(root_dir),
+        sources_dir=sim_struct.get_sources_dir(root_dir),
+        runs_dir=runs_dir,
+        fault_selection=None,
+        root_yaml=None,
+        output=None,
+        verbose=False,
+        models_dir=None,
+    )
+    df = est_cybershake(est_args)
+    grouped_df = df.groupby("fault_name").sum()
+    grouped_df.sort_index(axis=0, level=0, inplace=True)
+
+    # Sanity check
+    if np.any(grouped_df.index.values != fault_names):
+        raise Exception(
+            "The fault names of the estimation results and directories "
+            "are not matching. These have to match, quitting!"
+        )
+
+    # Create progress dataframe
+    column_t = []
+    for proc_type in PROCESS_TYPES:
+        column_t.append((proc_type, EST_CORE_HOURS_COL))
+        column_t.append((proc_type, ACT_CORE_HOURS_COL))
+    column_t.append(("total", EST_CORE_HOURS_COL))
+    column_t.append(("total", ACT_CORE_HOURS_COL))
+
+    progress_df = pd.DataFrame(
+        index=grouped_df.index.values, columns=pd.MultiIndex.from_tuples(column_t)
+    )
+
+    # Populate progress dataframe with estimation data
+    total = np.nan
+    for proc_type in PROCESS_TYPES:
+        values = grouped_df[proc_type, const.MetadataField.core_hours.value]
+        progress_df[proc_type, EST_CORE_HOURS_COL] = values
+        total = values if total is np.nan else total + values
```

Comment:
would it be worth vectorisation or a sub function for repeated code?

Reply:
I hope you are happy with this commit 
https://github.com/ucgmsim/slurm_gm_workflow/pull/405/commits/b13aea4bcd6ad3b9353b50b2b208b62c66d48084